py/stream.c: Implemented truncate().

DavidEGrayson opened the pull request:
This fixes issue #4775.

DavidEGrayson committed:
py/stream.c: Implemented truncate().

Signed-off-by: David Grayson <davidegrayson@gmail.com>

dpgeorge commented:
Thanks for the contribution.

As you can see above from the code size report, it's quite a large increase in code size.  More than I would have expected since the change is not that large.  Do you think there's any way to optimise further for size?

DavidEGrayson commented:
The majority of the code size change is due to the firmware now including file
truncation library routines for FAT and LFS which were previously
garbage-collected by the linker as unused functions.

Here are the results of some experiments where I measured the size of the RP2
port firmware with FAT and MSC enabled.  For the third row,
I added `return 0;` at the top of f_truncate, and for the fourth row, I added
that line to both f_truncate and lfs2_file_truncate.

|                 | RP2+FAT+MSC size | Change from above |
|-----------------|------------------|-------------------|
|Commit 3229791   | 323396           |                   |
|This PR, ec18976 | 323956           | +560              |
|empty f_truncate | 323756           | -200              |
|empty lfs2_file_truncate | 323580   | -176              |

So 67% of the code space increase comes from those library routines.

Perhaps truncate support should be an optional feature, and by default only
be enabled if `MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES` is 1.

Perhaps the code size could be improved a little by changing `f_truncate`
to take the new file size as an argument instead of just assuming we
want to truncate at the stream position.  The place where we call `f_truncate`
would get much simpler, and several references to `fp->fptr` inside the
function would get replaced by a local variable.

By the way, this PR currently uses f_lseek to expand a FAT file, and it
seems like that results in unspecified bytes being added to the end of the file.
If we improve f_truncate, we could make sure that only zeros are added,
which would be better.

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.
